datafusion/physical-expr/src/scalar_function.rs

```diff
@@ -142,25 +137,21 @@ impl PhysicalExpr for ScalarFunctionExpr {
     }
 
     fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
-        // evaluate the arguments, if there are no arguments we'll instead pass in a null array
-        // indicating the batch size (as a convention)
-        let inputs = match self.args.is_empty() {
-            // If the function supports zero argument, we pass in a null array indicating the batch size.
-            // This is for user-defined functions.
-            // MakeArray support zero argument but has the different behavior from the array with one null.
-            true if self.supports_zero_argument && self.name != "make_array" => {
-                vec![ColumnarValue::create_null_array(batch.num_rows())]
-            }
-            _ => self
-                .args
-                .iter()
-                .map(|e| e.evaluate(batch))
-                .collect::<Result<Vec<_>>>()?,
-        };
+        let inputs = self
+            .args
+            .iter()
+            .map(|e| e.evaluate(batch))
+            .collect::<Result<Vec<_>>>()?;
 
         // evaluate the function
         match self.fun {
-            ScalarFunctionDefinition::UDF(ref fun) => fun.invoke(&inputs),
+            ScalarFunctionDefinition::UDF(ref fun) => {
+                if fun.support_randomness() {
+                    fun.invoke_no_args(batch.num_rows())
```

Comment:
@jayzhan211 @alamb 
in order to resolve no arg function issue, we introduce two interface `support_randomness` and `invoke_no_args` for `ScalarUDFImpl` , I have no good idea to resolve the no arg function issue.

But it really increases barrier to implement a new  scalarUDF function, we need to know the attribute of the  new function.

If the attribute of the function is random, we must implement the interface of `nvoke_no_args`.

Reply:
Do we have any method to hide the special behavior?

Reply:
I don't quite understand `support_randomness` was needed. I gave it a try locally and I think we can avoid `support_randomness` like this: https://github.com/jayzhan211/arrow-datafusion/pull/2

Let me know what you think

Reply:
Functions that include randomness don't imply 0 args at all.  I think it is better to call `invoke_no_args` if it is no args 💯

Reply:
I forgot that `make_array` accept any number of args including 0, so we need to somehow specialize the function